CVS-46606 Custom node implementation for OCR (#541)
openvinotoolkit/model_server: docs/custom_node_development.md

# Custom Node Development Guide

## Overview

Custom node in OpenVINO Model Server simplifies linking deep learning models into a complete pipelines even if the inputs and output
of the sequential models do not fit. In many cases, output of one model can not be directly passed to another one.
The data might need to be analysed, filtered or converted to different format. Those operations can not be easily implemented
in AI frameworks or are simply not supported. Custom node addresses this challenge. They allows employing a dynamic library

> developed in C++ or C to perform arbitrary data transformations. 

Reviewer: While custom nodes were created with the purpose of simplifying the connections between models in DAG it can actually do more if one wishes since user can write basically any implementation there and just reuse our grpc/rest frontend (one cons is that such custom node execution is single threaded). 

@dtrawins / @dkalinowski I wonder if we want to mention it here as well?

Author: custom node execution could be multithreaded is someone decides to implement it that way. Anyway I think it is not the key use case to create standalone custom nodes without using OV and AI models at all.

## Custom Node API

The custom node library must implement the API interface defined in [custom_node_interface.h](../src/custom_node_interface.h).
The interface is defined in `C` to simplify compatibility with various compilers. The library could use third party components

> linked statically or dynamically. OpenCV is a built in component in OVMS which could be used to perform manipulation on the image

Reviewer: `linked statically or dynamically' this means that we have to test both cases

Author: we don't have tests for static linking but it should be doable if the namespace is not conflicting.

reports error, the pipeline execution is stopped and error is returned to the user. 

### "getInputsInfo" function
This function returns information about the metadata of the expected inputs. Returned CustomNodeTensorInfo object is used 
to create a response for getModelMetadata calls. It is also used in the user request validation and the pipeline 
configuration validation.

Custom nodes can generate the results which have dynamic size depending on the input data and the custom node parameters.
In such case function getInputsInfo should return value `0` on the dimension with dynamic size. It could be an input with
variable resolution or a batch size. 

### "getOutputInfo" function
Similar to previous function but defining the outputs metadata.

### "release" function
This function is called by OVMS at the end of the pipeline processing. It clear all memory allocations used during the 
node execution. This function should only call `free`. OVMS decides when to free and what to free.

## Using OpenCV
The custom node library can use any third-party dependencies which could be linked statically or dynamically.

> For simplicity there could be used OpenCV libraries which are already included in the OVMS docker image.

Reviewer: ```suggestion
For simplicity OpenCV libraries included in the OVMS docker image can be used.
```

Just add include statement like:
```c++
#include "opencv2/core.hpp"
```

## Building

> Custom node library can be compiled using any tool but for simplicity there are prepared a method based on a g++ compiled 

Reviewer: Cannot understand this sentence

inside a docker container. A simple example is described in this [Makefile](../src/custom_nodes/east_ocr/Makefile). 

## Testing 
Recommended method for tesing the custom library is via OVMS execution:

> - Compile the library using a docker container like configured in the Makefile. It will be exported to `lib` folder.

Reviewer: Unnecessary "like"?

Reviewer: Add full path instead of 'lib' folder?

- Prepare a pipeline configuration with the path custom node library compiled in the previous step
- Start OVMS docker container
- Submit a request to OVMS endpoint using a gRPC or REST client
- Analyse the logs on the OVMS server

For debugging steps refer the OVMS [developer guide](developer_guide.md)

## Custom node examples 
The best staring point for developing new custom nodes is via exploring or copying the examples.

The fully functional custom nodes are in:
- [east-resnet50 OCR custom node](../src/custom_nodes/east_ocr)

> Other examples are included in the unit tests:

Reviewer: I would limit ourselves to mentioning the path to the custom nodes and 2/3 libraries that do something with positive results. I don;t think that mentioning mocks here is useful

> - [node_add.c](../src/test/custom_nodes/node_add_sub.c)

Reviewer: node_add mentioned 2 times & it should be add_sub I think

- [node_choose_maximum.cpp](../src/test/custom_nodes/node_choose_maximum.cpp)

> - [node_add.c](../src/test/custom_nodes/node_missing_implementation.c)

Reviewer: missing_implementation
